### ISSUE TITLE
Create a Cyclone DX SBOM equivalent to the current build build-info json metadata

### ISSUE THREAD

**Issue by andrew-m-leonard** [adoptium/temurin-build#2753]
As part of the strategy to supply Cyclone DX SBOM info for Adoptium binaries, this issue covers the initial task of creating a Cyclone DX SBOM equivalent of the current binary metadata build-info.
This metadata can be seen here as an example: https://github.com/adoptium/temurin17-binaries/releases/download/jdk-17%2B35/OpenJDK17-jdk_x64_linux_hotspot_17_35.tar.gz.json
To view these files it is best to download and view using a suitable JSON viewer. (you can download, and open in a browser).
This JSON contains numerous pieces of "build-info",eg:
- "version" : Version info for the binary, eg.what openjdk tag, build, ..
- scmRef: What adoptium openjdk mirror git tag it was built from
- buildRef : What level (git commit SHA) of temurin-build scripts were used to build it
- full_version_output: the built binary "java -version" string
- makejdk_any_platform_args: the temurin build make arguments used
- configure_arguments: the openjdk "Configure" arguments used
- make_command_args: the openjdk make arguments used
- openjdk_built_config: the Temurin build scripts configuration generated
- openjdk_source: the openjdk repo source that was built (git SHA)
- build_env_docker_image_digest: if built within a docker container, this is the exact SHA digest of the image it was built from
- dependency_version_alsa|freetype|freemarker: various dependencies built with, URLs to the version used

The intention of this task is to take this current build-info and generate a CycloneDX SBOM for it.

**Comment by andrew-m-leonard** [comment]
@SehrishHussain fyi

**Comment by SehrishHussain** [comment]
@andrew-m-leonard my question was/is still the same i.e. how/where do we generate these CycloneDX SBoM? I have been searching over the internet and asked some relevant people. Meanwhile, I want to work on another task in parallel. What do you suggest please.

**Comment by andrew-m-leonard** [comment]
@SehrishHussain great work Sehrish, that looks a great first schema.
So agree automation through one of the available plugins would be the desired way, but as you can see what we have is rather bespoke at the moment. 
We need to break down the current build-info, into constituent components, some info is "version/input" info, some is "how it was made" info (Assembly), and some is "3rd party dependencies". 
The BUILT_CONFIG info from the build-info for example needs better organization and maybe bits of it categorizing:
https://github.com/adoptium/temurin17-binaries/releases/download/jdk-17%2B35/OpenJDK17-jdk_x64_linux_hotspot_17_35.tar.gz.json

**Comment by SehrishHussain** [comment]
@andrew-m-leonard I've created first automated SBOM. It's very basic but its automated. Used CycloneDX CLI tool. I need to build Temurin locally, for some unkown reason my machine isn't building. Anyhow, I thought to share it here. I will get a new machine with Ubuntu and hopefully will build jdk binaries locally as it is the first step for automating SBOM.

**Comment by SehrishHussain** [comment]
@andrew-m-leonard I've created first automated SBOM. It's very basic but its automated. Used CycloneDX CLI tool and above Json file. I need to build Temurin locally, for some unkown reason my machine isn't building. Anyhow, I thought to share it here. I will get a new machine with Ubuntu and hopefully will build jdk binaries locally as it is the first step for automating SBOM. Its a JSON file so can't attach it here. Here is the content of file. 
`{
  "bomFormat": "CycloneDX",
  "specVersion": "1.3",
  "serialNumber": "urn:uuid:a29e03ec-6cdb-49d3-899e-3263a6d04486",
  "metadata": {
    "timestamp": "2021-11-01T17:46:46Z",
    "tools": [
      {
        "vendor": "CycloneDX",
        "name": "CycloneDX CLI",
        "version": "0.19.0.0"
      }
    ]
  },
  "components": []
}`

**Comment by andrew-m-leonard** [comment]
That's great @SehrishHussain , let me know if you need help getting your local build working.
It would perhaps be nice to do a quick 2min demo?

**Comment by SehrishHussain** [comment]
@andrew-m-leonard yes please, that would be very helpful.

**Comment by andrew-m-leonard** [comment]
@SehrishHussain the above CycloneDX JSON looks good. What we need to do next is get that produced as an output of the build. That probably means something along the lines of the following:
- Add CycloneDX tooling to build infrastructure...so it is available at least on certain nodes to build the SBOM. What tools need to be available?
- Design how you will invoke and gather the current metadata into the Cyclone DX SBOM...?
- Info on the current metadata is here: https://github.com/adoptium/temurin-build/blob/master/README.md#metadata
- Code is here for current metadata construction: https://github.com/adoptium/ci-jenkins-pipelines/blob/9556d1f74dfe9ad5cb2dcf14df33468a171533f0/pipelines/build/common/openjdk_build_pipeline.groovy#L715
- Actual writing of the Metadata JSON is here: https://github.com/adoptium/ci-jenkins-pipelines/blob/9556d1f74dfe9ad5cb2dcf14df33468a171533f0/pipelines/build/common/openjdk_build_pipeline.groovy#L921
- Metadata groovy class : https://github.com/adoptium/ci-jenkins-pipelines/blob/b9b8f47ae6f6828734ede7a4869eee8bae8c6505/pipelines/library/src/common/MetaData.groovy

Feel free to discuss your ideas and designs here..
thanks

**Comment by andrew-m-leonard** [comment]
CycloneDX core-java tool provides Bom json generators which we can use to call from the Jenkins pipline:
https://cyclonedx.github.io/cyclonedx-core-java/org/cyclonedx/generators/json/BomJsonGenerator13.html

The pipeline write metadata code here: https://github.com/adoptium/ci-jenkins-pipelines/blob/9556d1f74dfe9ad5cb2dcf14df33468a171533f0/pipelines/build/common/openjdk_build_pipeline.groovy#L921
can call a Java app using context.sh(script: "java AdoptiumGenSBOM ...")
The provisor to this that we ensure a JDK is available...

The Java app AdoptiumGenSBOM, could be have a command line interface like:
```
java AdoptiumBom --create bom.json --name Temurin_x64Linux --version jdk-18+26
java AdoptiumBom --update bom.json --component freetype --version 1.18 --hash 24abcdef4545
```

**Comment by andrew-m-leonard** [comment]
@SehrishHussain the PR for the CycloneDX jenkins build integration is now merged: https://github.com/adoptium/temurin-build/pull/2805

To allow you to extend and test changes/additions to TemurinGenSBOM.java, you can run it as follows:
**Prereqs locally:**
- Install a suitable Java, eg.from here: https://adoptium.net/releases.html
- Ensure "ant" installed

1.  Clone temurin-build: git clone git@github.com:adoptium/temurin-build.git
2. cd temurin-build/cyclonedx-lib
3. ant -f build.xml clean
4. ant -f build.xml build
5. ant -f build.xml run

build.xml "run" target just invokes TemurinGenSBOM with no args, you can change that to test whatever you need..

Next step is to "design" how we transform/gather the temurin build-info to call this app, so look at what "build-info" we currently have, and how that would be represented in the SBOM? The design a command line to achieve this, example:
```
java TemurinGenSBOM --create bom.json --name Temurin_x64Linux --version jdk-18+26 --type application
java TemurinGenSBOM --add_dependency bom.json --name freetype --version 1.18 --hash 24abcdef4545
...
```

**Comment by andrew-m-leonard** [comment]
This is now complete, and can be enabled by specifying a buildArg of --create-sbom
